apollo-router/src/spec/schema.rs

```diff
@@ -470,6 +553,230 @@ impl Schema {
+    // TODO: do something with warnings and advice?
+    errors.retain(|diagnostic| diagnostic.is_error());
+    if !errors.is_empty() {
+        return Err(SchemaError::Validation(errors));
+    }
+
+    // Trace log recursion limit data
+    let recursion_limit = compiler.db.ast().recursion_limit();
+    tracing::trace!(?recursion_limit, "recursion limit data");
+
+    fn as_string(value: &hir::Value) -> Option<&String> {
+        if let hir::Value::String(string) = value {
+            Some(string)
+        } else {
+            None
+        }
+    }
+
+    let mut subgraphs = HashMap::new();
+    // TODO: error if not found?
+    if let Some(join_enum) = compiler.db.find_enum_by_name("join__Graph".into()) {
+        for (name, url) in join_enum
+            .enum_values_definition()
+            .iter()
+            .filter_map(|value| {
+                let join_directive = value
+                    .directives()
+                    .iter()
+                    .find(|directive| directive.name() == "join__graph")?;
+                let name = as_string(join_directive.argument_by_name("name")?)?;
+                let url = as_string(join_directive.argument_by_name("url")?)?;
+                Some((name, url))
+            })
+        {
+            if url.is_empty() {
+                return Err(SchemaError::MissingSubgraphUrl(name.clone()));
+            }
+            let url = Uri::from_str(url).map_err(|err| SchemaError::UrlParse(name.clone(), err))?;
+            if subgraphs.insert(name.clone(), url).is_some() {
+                return Err(SchemaError::Api(format!(
+                    "must not have several subgraphs with same name '{}'",
+                    name
+                )));
+            }
+        }
+    }
+
+    let object_field = |field: &hir::FieldDefinition| (field.name().to_owned(), field.ty().into());
+    let implements_interface = |imp: &hir::ImplementsInterface| imp.interface().to_owned();
+    let mut object_types: HashMap<_, _> = compiler
+        .db
+        .object_types()
+        .iter()
+        .map(|def| {
+            (
+                def.name().to_owned(),
+                ObjectType {
+                    name: def.name().to_owned(),
+                    fields: def.fields_definition().iter().map(object_field).collect(),
+                    interfaces: def
+                        .implements_interfaces()
+                        .iter()
+                        .map(implements_interface)
+                        .collect(),
+                },
+            )
+        })
+        .collect();
+    for ext in compiler.db.object_type_extensions().iter() {
+        if let Some(def) = object_types.get_mut(ext.name()) {
+            def.fields
+                .extend(ext.fields_definition().iter().map(object_field));
+            def.interfaces
+                .extend(ext.implements_interfaces().iter().map(implements_interface))
+        } else {
+            failfast_debug!(
```

Comment:
would that be an issue in apollo-compiler, or in the schema? If it's an invalid schema, that should not warrant a panic

Reply:
This macro only panics when the router is compiled with `--features failfast`. I made this code use it because the corresponding existing code uses it (in the function now renamed `parse_with_apollo_parser`)

Extending an undefined object type is a schema error. Soon this logic will be in apollo-compiler which will emit a diagnostic for such errors.